File: Data Structures & Algorithms/trapping-rain-water/submission-0.java
class Solution {
    public int trap(int[] height) {
        int maxWater = 0, water = 0;
        int l = 0, r = height.length - 1;
        int leftMax = height[l], rightMax = height[r];

        while (l < r){
            if (leftMax < rightMax){
                l += 1;
                leftMax = Math.max(leftMax, height[l]);
                water = leftMax - height[l];
            } else {
                r -= 1;
                rightMax = Math.max(rightMax, height[r]);
                water = rightMax - height[r];
            }
            maxWater += water;
        }
        return maxWater;
    }
}

File: Data Structures & Algorithms/max-water-container/submission-0.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int i = 0, j = heights.size() - 1;
        int maxWater = 0;
        while (i < j) {
            if (heights[i] < heights[j]) {
                int water = (j - i) * heights[i];
                maxWater = std::max(maxWater, water);
                i++;
            }
            else {
                int water = (j - i) * heights[j];
                maxWater = std::max(maxWater, water);
                j--;
            }
            // std::cout << maxWater << "\n";
        }
        return maxWater;
    }
};
